### PR TITLE
replace json path lib

### PR DESCRIPTION
**Pull request by rickardenglund**

### CODE REVIEW

File: api/godog/json/matcher.go [hunk 1]
```diff
@@ -55,270 +52,76 @@ func Match(json []byte, path string, pattern string) error {
...
-}
-
-func ReadStringArr(json []byte, path string) (result []string, err error) {
-	t, err := resolve(json, path)
-	if err != nil {
-		return []string{}, err
-	}
+func revertLegacySyntax(path string) string {
+	// Remove brackets earlier they were used for array indexing
+	bracketRegexp := regexp.MustCompile(`\.?\[(\d+)\]`)
```
Comment (j0s) [inline api/godog/json/matcher.go:68]: Why `\.?` in the beginning of the regexp? An optional character before the brackets?

**Review by j0s — COMMENTED** [review]
Really nice to get rid of all this code!

File: api/godog/json/matcher.go (hunk 1 above)
Comment (rickardenglund) [reply api/godog/json/matcher.go:68]: Added a test case to show where the initial dot is needed.

in case of path = .[0]

Comment (j0s) [reply api/godog/json/matcher.go]: I see, `\.` matches an actual dot :+1:

File: api/godog/json/matcher.go [hunk 2]
```diff
@@ -55,270 +30,76 @@ func Match(json []byte, path string, pattern string) error {
...
+	// Remove brackets earlier they were used for array indexing
+	bracketRegexp := regexp.MustCompile(`\.?\[(\d+)]`)
+	path = bracketRegexp.ReplaceAllString(path, ".$1")
 
-	segments, err := parseMatchSegments(path)
-	if err != nil {
-		return t, err
+	if strings.HasPrefix(path, ".") {
+		path = string([]rune(path)[1:])
 	}
```
Comment (j0s) [inline api/godog/json/matcher.go:51]:
```suggestion
	path = strings.TrimPrefix(path, ".")
```

File: api/godog/json/matcher.go [hunk 3]
```diff
@@ -55,270 +30,76 @@ func Match(json []byte, path string, pattern string) error {
...
+	path = revertLegacySyntax(path)
 
-func encodeString(lex *scanner.Scanner) (term, error) {
-	str, err := strconv.Unquote(lex.TokenText())
+	res := gjson.Get(string(json), path)
 
-	if err != nil {
-		return null, err
+	if !res.Exists() || res.IsObject() || res.IsArray() {
+		return "", errors.Errorf("Match error: Expected a scalar got '%T' JSON: %s", res.String(), string(json))
```
Comment (j0s) [inline api/godog/json/matcher.go:79]: Isn't `res.String()` printed as `%T` always just string?

File: api/godog/json/matcher.go [hunk 4]
```diff
@@ -55,270 +30,76 @@ func Match(json []byte, path string, pattern string) error {
...
-		}
-
-		arr = append(arr, val)
-
-		if lex.Peek() == ',' {
-			lex.Scan()
+	for _, v := range res.Array() {
+		stringValue := v.String()
+		if !(v.IsObject() || !v.IsArray()) {
+			return []string{}, errors.Errorf("Match error: Expected a scalar got: %s", v.String())
```
Comment (j0s) [inline api/godog/json/matcher.go:99]: You can use `stringValue` here instead of calling `v.String()` again

File: api/godog/json/matcher.go [hunk 5]
```diff
@@ -55,270 +30,76 @@ func Match(json []byte, path string, pattern string) error {
...
-			return null, err
-		}
-
-		arr = append(arr, val)
-
-		if lex.Peek() == ',' {
-			lex.Scan()
+	for _, v := range res.Array() {
+		stringValue := v.String()
+		if !(v.IsObject() || !v.IsArray()) {
```
Comment (j0s) [inline api/godog/json/matcher.go:98]: I don't understand this; if something is not an object but an array, we throw an error.

If it's neither an object nor an array then I guess it's a scalar so then I understand it's OK. But if v is an object it's OK even if it's also an array. Should non-scalar objects be accepted at all? Can something can be both an object and an array?

If the logic is correct, maybe a small comment could be useful? Could it maybe be made a little bit clearer by doing  this instead:
```suggestion
		if !v.IsObject() && v.IsArray()) {
```

File: api/godog/json/matcher.go [hunk 6]
```diff
@@ -2,51 +2,26 @@ package json
...
+	result := gjson.Get(string(json), path)
 
-	if t.value != nil {
-		return errors.Errorf("Match error: Expected null got '%T' JSON: %s", t.value, string(json))
+	if result.Exists() && !result.IsArray() && !result.IsObject() && result.Value() == nil {
+		return nil
 	}
 
-	return nil
+	return errors.Errorf("Match error: Expected null got '%T' JSON: %s", result.String(), string(json))
```
Comment (j0s) [inline api/godog/json/matcher.go:24]: Isn't `result.String()` printed as `%T` always just string?

**Review by j0s — COMMENTED** [review]
Some small comments

File: api/godog/json/matcher.go (hunk 3 above)
Comment (rickardenglund) [reply api/godog/json/matcher.go:68]: indeed :)

File: api/godog/json/matcher.go (hunk 2 above)
Comment (rickardenglund) [reply api/godog/json/matcher.go]: nice :)

File: api/godog/json/matcher.go (hunk 5 above)
Comment (rickardenglund) [reply api/godog/json/matcher.go]: Yes, this was wrong

**Review by j0s — APPROVED** [review]

**Pull request merged**